maven-core/src/main/java/org/apache/maven/DefaultMaven.java

```diff
@@ -367,11 +354,37 @@ private MavenExecutionResult doExecute( MavenExecutionRequest request, MavenSess
         return result;
     }
 
-    private void afterSessionStart( MavenSession session )
+    private void finishRepositorySystemSession( MavenSession session, DefaultRepositorySystemSession repoSession )
+        throws ComponentLookupException
+    {
+        WorkspaceReader reactorWorkspace = container.lookup( WorkspaceReader.class, ReactorReader.HINT );
+        //
+        // Desired order of precedence for local artifact repositories
+        //
+        // Reactor
+        // Workspace
+        // User Local Repository
+        //
+
+        // we only lookup the project scoped ones here, lookup at system level will also find the system components
+        // (like ReactorReader) but that does not work well as these are special handled!
+        for ( WorkspaceReader workspaceReader : getProjectScopedExtensionComponents( session.getProjects(),
+                                                                                     WorkspaceReader.class ) )
+        {
+            repoSession.setWorkspaceReader( ChainedWorkspaceReader.newInstance( repoSession.getWorkspaceReader(),
```

Comment:
If there are multiple `WorkspaceReader` implementations, this would create a chain that is larger than necessary; each new extension would create a new `ChainedWorkspaceReader` that points to the previous `ChainedWorkspaceReader` and adds the one it just detected in the extension.

Wouldn't it make more sense to first find all the extension components and then build only one `ChainedWorkspaceReader` with all `WorkspaceReaders` found in all the extensions?

Reply:
Sure that is possible, I just adopted the "style" currently used, but ChainedWorkspaceReader accepts also an array,  I can  adjust ChainedWorkspaceReader to combine if any of the provided ones is already a chained one. Let me know if you think this should be done.

Reply:
Okay, ChainedWorkspaceReader is not part of this repository should I suggest a change in ChainedWorkspaceReader or should I handle it in MavenDefault?

Reply:
done